src/lib/auth/authNZ.mjs

```diff
@@ -0,0 +1,192 @@
+  const accountCreds = container.resolve("accountCreds");
+  try {
+    const creds = accountCreds.get({ key: profile });
+    return creds.account_key;
+  } catch (e) {
+    if (e instanceof CredsNotFoundError) {
+      // Throw InvalidCredsError back up to middleware entrypoint to prompt login
+      throw new InvalidCredsError();
+    }
+    e.message = `Error getting account key for ${profile}: ${e.message}`;
+    throw e;
+  }
+}
+
+async function checkAccountKeyRemote(accountKey) {
+  const accountClient = container.resolve("accountClient");
+  // If account key is invalid or expired, this will throw InvalidCredsError
+  try {
+    return await accountClient.whoAmI(accountKey);
+  } catch (e) {
+    if (e instanceof InvalidCredsError) {
+      // Return null to indicate account key is invalid and we will try to refresh it
+      return null;
+    }
+    throw e;
+  }
+}
+
+async function refreshSession(profile) {
+  const accountClient = container.resolve("accountClient");
+  const accountCreds = container.resolve("accountCreds");
+  const creds = accountCreds.get({ key: profile });
+  const { refresh_token } = creds;
+  if (!refresh_token) {
+    throw new Error(
+      "Invalid access_keys file configuration for profile: " + profile,
+    );
+  }
+  // If refresh token expired, this will throw InvalidCredsError
+  const newCreds = await accountClient.refreshSession(refresh_token);
+  return newCreds;
+}
+
+async function setDBKey({ accountKey, path, role, url }) {
+  const secretCreds = container.resolve("secretCreds");
+  const accountClient = container.resolve("accountClient");
+  const existingSecret = getDBKey({ accountKey, path, role });
+  if (existingSecret) {
+    // If this throws an error, user
+    const dbKeyIsValid = await checkDBKeyRemote(existingSecret.secret, url);
+    if (dbKeyIsValid) {
+      return existingSecret;
+    }
+  }
+  // DB key doesn't exist locally, or it's invalid. Create a new one, overwriting the old
+  const newSecret = await accountClient.createKey({
+    accountKey,
+    path,
+    role,
+  });
+  secretCreds.save({
+    creds: {
+      path,
+      role,
+      secret: newSecret.secret,
+    },
+    key: accountKey,
+  });
+  return newSecret;
+}
+
+export function getDBKey({ accountKey, path, role }) {
+  const secretCreds = container.resolve("secretCreds");
+  try {
+    const existingCreds = secretCreds.get({ key: accountKey, path, role });
+    // TODO: type here is the same format returned from FD createKey
+    return {
+      secret: existingCreds,
+      path,
+      role,
+    };
+  } catch (e) {
+    if (e instanceof CredsNotFoundError) {
+      return null;
+    }
+    e.message = `Error getting secret for ${accountKey} ${path} ${role}: ${e.message}`;
+    throw e;
+  }
+}
+
+export async function checkDBKeyRemote(dbKey, url) {
+  const client = new FaunaClient({
+    secret: dbKey,
+    endpoint: url,
+  });
+  const result = await client.query("0");
```

Comment:
is there any advantage to pre-checking instead of trying to do the work? is it that we can do it still in the middleware before reaching the handler?